### ISSUE TITLE
Deep pattern matching in kernel language

### ISSUE THREAD

**Issue by david-christiansen** [gelisam/klister#107]
After working a bit with pattern macros, I think we should have deep pattern matching in the core language rather than as a library.

This is mainly because it makes pattern macros much more useful. Right now, we can't define e.g. `list` as a pattern macro, because it won't work here. I don't think it's so hard to implement, either, especially if we're not caring too much about optimal performance (although proper case trees are not that hard to do either, and probably worth it).

Another thing to consider is whether pattern macros should actually expand to patterns, or whether we should expose some sort of more fundamental interface (e.g. success/failure continuations), which would enable things like view patterns or guards to be defined, I think. But that's for another time, I suppose.

**Comment by langston-barrett** [comment]
TBQH I know very little about the implementation of pattern matching, so don't feel especially qualified to comment on good points in the design space or amounts of effort any changes would take.

**Comment by gelisam** [comment]
Just to make sure I understand: `(list 1 2)` would be a pattern which matches on `(cons 1 (cons 2 nil))`, and while it seems like it should be possible to expand `(match scrutinee [(list 1 2) rhs])` to

```
(match scrutinee
  [(cons 1 xs)
   (match xs
     [(cons 2 ys)
      (match ys
        [nil rhs])])])
```

This would have to be done by defining a special `match` rather than a special `list`, because `(list 1 2)` can only expands to something which fills the `...` in `(match scrutinee [... rhs])`, it cannot cause the code around it to reshape itself to (match scrutinee [... (match ...)])`.

I agree that this is limitation is quite upsetting, and that it's worth changing the language to make it possible to implement the `list` pattern as a macro. However, given the minimalist core + user macros nature of our language so far, I like the idea of exposing a more fundamental API much better! Shall we open another ticket to discuss what that might look like, or shall we discuss here?

**Comment by gelisam** [comment]
Here's one idea. Instead of matching on `(list 1 2 3)`, pattern macros could be given the entire `match` expression:

```
(define-macro (list stx)
  (syntax-case stx
    [(match scrutinee [(list) rhs] cases ...)
     `(match ,scrutinee [(nil) ,rhs] ,@cases) 
    [(match scrutinee [(list x xs ...) rhs] cases ...)
     `(match ,scrutinee
        [(cons ,x tail)
         (match tail
           [(list ,@xs) ,rhs]
           ,@cases)]
        ,@cases)]))
```

There are a few details to iron out such as patterns like `(cons (list 1 2 3) xs)` and the exponential multiplication of the `,@cases`, but I don't think those are too difficult to fix.

The thing I like with that approach is that it suggests a generalization: extensible macros. An extensible macro `ext` is one which, when pattern-matching on its input Syntax, has some extension points in mind, detects macro names at those extension points, and calls that extension macro `helper` with a Syntax object which might not have the shape `(helper args ...)` but some other shape agreed upon by convention, most commonly `(ext (helper helper-args ...) ext-args ...)`. This way, pattern macros don't have to be special, users can write their own extensible macros!

In fact, expression macros don't have to be special either, they could be extension macros which expect the shape `(helper helper-args ...)`, or even `(#%app helper args ...)`. Then definition macros can use `(#%module (helper helper-args ...) module-args ...)`, etc., and now no kind of macro needs to be special. Macros which want to work in both expression, pattern, and definition context could look at the shape of the syntax they are given and behave accordingly.

This implies a completely different mechanism for asking which problem the macro is currently solving. In the expression case, we thus need a way for the macro to ask what type the caller expects its generated expression to have. Well, instead of `(#%app helper args ...)`, how about `(the (helper args ...) type)`? That is, perhaps a Syntax should be allowed to have some pieces which are not yet known, e.g. because another macro needs to be run in order to obtain that piece, or because that piece is filled-in by the type-checker. Then pattern-matching on that piece of Syntax causes the macro to get stuck, as usual.

The biggest downside of this last idea, I think, is that pattern-matching on a Syntax is now a side-effect instead of being pure. But the biggest upside, I think, is that it opens the way for extensible macros which define their own idea of what a helper macro's goal is, and which provide their extension macros with that goal as part of the input Syntax. It should even be possible for that extensible macro to define goals which, like types, start as partially-unknown and get more known over time as the extension macros contribute more information and the extensible macro integrates that information. We just need a syntax for distinguishing whether`(helper (generate-goal))` means that `helper` receives the Syntax `(generate-goal)` or a partial Syntax which is stuck until `generate-goal` generates a goal Syntax.

**Comment by gelisam** [comment]
Come to think of it, having macros cooperate via Syntax values and convention feels rather untyped. What if we generalized extension macros from being functions of type `Syntax -> Macro Syntax` to just being values of arbitrary types which happen to be known at compile-time? Then, instead of having `match` look for a `Syntax -> Macro Syntax` at its extension points and pass it `(match scrutinee [(helper ...)] ...)`, it can instead look for a `Syntax -> Syntax -> Macro Syntax`and pass it success and failure continuations. And `#%app` can look for a `Syntax -> Type -> Macro Syntax` and pass it the usual input Syntax and the partial type which the macro can get stuck on.

How would a macro now express that it wants to work both as a pattern macro and an expression macro? We'd need to somehow attach two values of different types to the same identifier. And we can! Well, we will soon. We are currently thinking about global monotonic stores to which arbitrary macros could contribute, right? And a store may map identifiers to values of a specific type, e.g. a number indicating the arity of a type? Well, similarly, we could have a store which maps identifiers to `Syntax -> Syntax -> Macro Syntax` values, and a store which maps identifiers to `Syntax -> Type -> Macro Syntax` values, and we can simply bind the same identifier in both stores.

The biggest problem with that approach is that those stores use threshold reads, and thus an extensible macro would not be able to lookup an identifier in that store and determine whether that identifier maps to an extension macro or not; in the "not" case, the extensible macro would get stuck forever. I guess that just means we need two kinds of stores, one which arbitrary macros can contribute to and thus need threshold reads, and one which relies on scoping rules to guarantee that if we don't have a value for a given identifier yet, we never will, and thus we don't need to block?

**Comment by david-christiansen** [comment]
> Just to make sure I understand: `(list 1 2)` would be a pattern which matches on `(cons 1 (cons 2 nil))`, and while it seems like it should be possible to expand `(match scrutinee [(list 1 2) rhs])` to
> 
> ```
> (match scrutinee
>   [(cons 1 xs)
>    (match xs
>      [(cons 2 ys)
>       (match ys
>         [nil rhs])])])
> ```
> 
> This would have to be done by defining a special `match` rather than a special `list`, because `(list 1 2)` can only expands to something which fills the `...` in `(match scrutinee [... rhs])`, it cannot cause the code around it to reshape itself to `(match scrutinee [... (match ...)])`.

The problem is worse. What if we had this?
```
(case whatever
  [(pair (list 1 2 3) (list 3 2 1)) RHS])
```
Rewriting this into nested `case` expressions would require that `case` was a fair bit more complicated. We can implement it either as a traditional compiler into case trees à la Augustsson (which is what Racket's does), but this substantially complicates the implementation of macros in pattern positions. (As an aside, Ballentyne, King, and Felleisen's upcoming ECOOP paper has good stuff we should implement for extensible DSLs, but that's a bit beyond the scope of what we put in our core language for pattern matching.) We can also do the simpler two-continuation translation that we have right now. In either case, pattern macros become quite complicated to implement.

> I agree that this is limitation is quite upsetting, and that it's worth changing the language to make it possible to implement the `list` pattern as a macro. However, given the minimalist core + user macros nature of our language so far, I like the idea of exposing a more fundamental API much better! Shall we open another ticket to discuss what that might look like, or shall we discuss here?

I personally thing we should just stick deep patterns in to the core language. It's not a hugely controversial feature for an ML-like language, it's fairly well-understood, and it simplifies lots of other things. Just as we don't have a core language where everything has to be written in ANF or CPS for expressions, I think it's OK to also have eliminations look like a normal ML.

**Comment by david-christiansen** [comment]
BTW, the translation of the above example would be something like this:
```
(let ((fk (lambda (_) (case whatever REST-OF-CASES))))
(case whatever
  [(pair x y)
   (case x
    [(:: x1 x*)
     (case x* [(:: x2 x**) (case x** [(:: x3 x***) (case x*** [(nil) (case y (... DO SAME THING FOR Y ... (if (= x1 1) (if (= x2 2) ....) RHS] [_ (fk unit)]) [_ (fk unit)]) [_ (fk unit)]) ...)]))
```

Not very fun ...

**Comment by david-christiansen** [comment]
I'm also not a big fan of the compiler generating syntax objects to communicate aspects of the core language, nor of letting macros examine the context in which they're expanded.

In the first case, if the compiler has already checked a thing, then it should use a datatype that precisely describes the thing, rather than just quoting it up to some arbitrary syntax.

In the second case, macro expansion goes from the outside inwards. By the time the macro expander gets to an internal bit of syntax, the context that's immediately around it has been expanded to the core language. In other words, we have something like a zipper into a `Fix (Compose CoreF Maybe)`, or possibly a zipper into a `(Fix (Compose CoreF (Either Syntax)))` not a syntax object with holes in it.

OTOH, it sounds a bit like we're groping towards an understanding of extensible pattern matching as a kind of control operator delimited by the outermost case expression. This is an interesting way to go, and I'd want to read up and talk to some experts. But I also don't want the perfect to be the enemy of the good, and some other things that I'd like to simplify in the language (like making `syntax-case` both better and non-primitive) are currently waiting on something like this. Given that our core language is intended to be a fairly normal ML-ish language, is full pattern matching such an issue to add, potentially with the understanding that the macro API to it is subject to change?

**Comment by gelisam** [comment]
Oh, I'm not saying that I don't want to merge your deep pattern matching work, I'm only saying that I am excited about the future API we'd eventually replace it with, the "more fundamental interface (e.g. success/failure continuations), which would enable things like view patterns or guards to be defined", and so I'm sharing the ideas these words inspired me before I forget them. I should go review your PR now.

I was also not saying that the extension macros would get the ability to examine their context, I'm saying that extensible macros would call the `Macro -> Syntax Macro` function which defines that extension macro with a different Syntax input than our expander currently gives macros. It doesn't have to match the context in which the macro is called, I just think it's more intuitive for the writers of the extension macros that way, but the extensible macro could just as well decide to send e.g. a Syntax wrapping a single number, or a tuple of arguments, or whatever it wants.
